File: examples/15-virtual-memory/go/main.go
// memmap: self-inspecting virtual-memory tool — perform one kind of
// allocation, touch its pages, then report what the kernel now shows for
// this very process: the /proc/self/maps region backing the allocation,
// the VmRSS growth, and the getrusage(2) page-fault deltas.
//
//	memmap --mode stack|heap|mmap-anon|mmap-file <FILE>|fault-walk [--mb N]
//
// Output contract (identical across C++, Go, Rust):
//
//	memmap: mode=<m> bytes=<b> pages=<p>
//	[fault-walk only] memmap: walk file=<path> steps=8
//	[fault-walk only] memmap: step=<i>/8 pages=<p> minor=<d> major=<d>
//	memmap: maps excerpt
//	memmap:   <raw /proc/self/maps line>   <-- target (mode=<m>)
//	memmap: vmrss_before=<kb>KB vmrss_after=<kb>KB
//	memmap: faults minor=<n> major=<n>
//
// Go note: goroutine stacks are runtime-managed memory, not the kernel's
// [stack] mapping, so the stack-mode excerpt shows an anonymous region.
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"unsafe"

	"golang.org/x/sys/unix"
)

const (
	mib        = 1 << 20
	stackCapMb = 4 // stays well inside typical stack expectations
	walkSteps  = 8
)

// sink defeats dead-store elimination of the touch loops.
var sink byte

type options struct {
	mode string
	mb   int
	file string
}

type baseline struct {
	rssKB int64
	minor int64
	major int64
}

// ---------------------------------------------------------------------------
// /proc and getrusage probes
// ---------------------------------------------------------------------------

func vmrssKB() (int64, error) {
	f, err := os.Open("/proc/self/status")
	if err != nil {
		return 0, fmt.Errorf("open /proc/self/status: %w", err)
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := sc.Text()
		if !strings.HasPrefix(line, "VmRSS:") {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) < 2 {
			return 0, fmt.Errorf("cannot parse VmRSS: %q", line)
		}
		kb, err := strconv.ParseInt(fields[1], 10, 64)
		if err != nil {
			return 0, fmt.Errorf("cannot parse VmRSS: %w", err)
		}
		return kb, nil
	}
	if err := sc.Err(); err != nil {
		return 0, fmt.Errorf("read /proc/self/status: %w", err)
	}
	return 0, fmt.Errorf("VmRSS not found in /proc/self/status")
}

func faultCounts() (minor, major int64, err error) {
	var ru unix.Rusage
	if err := unix.Getrusage(unix.RUSAGE_SELF, &ru); err != nil {
		return 0, 0, fmt.Errorf("getrusage: %w", err)
	}
	return ru.Minflt, ru.Majflt, nil
}

func takeBaseline() (baseline, error) {
	rss, err := vmrssKB()
	if err != nil {
		return baseline{}, err
	}
	minor, major, err := faultCounts()
	if err != nil {
		return baseline{}, err
	}
	return baseline{rssKB: rss, minor: minor, major: major}, nil
}

// printMapsExcerpt prints every /proc/self/maps line overlapping
// [addr, addr+length).
func printMapsExcerpt(addr, length uintptr, mode string) error {
	f, err := os.Open("/proc/self/maps")
	if err != nil {
		return fmt.Errorf("open /proc/self/maps: %w", err)
	}
	defer f.Close()
	fmt.Println("memmap: maps excerpt")
	lo, hi := uint64(addr), uint64(addr+length)
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 64*1024), 64*1024)
	for sc.Scan() {
		line := sc.Text()
		rangePart, _, ok := strings.Cut(line, " ")
		if !ok {
			continue
		}
		startStr, endStr, ok := strings.Cut(rangePart, "-")
		if !ok {
			continue
		}
		start, err1 := strconv.ParseUint(startStr, 16, 64)
		end, err2 := strconv.ParseUint(endStr, 16, 64)
		if err1 != nil || err2 != nil {
			continue
		}
		if start < hi && end > lo {
			fmt.Printf("memmap:   %s   <-- target (mode=%s)\n", line, mode)
		}
	}
	return sc.Err()
}

// report prints the common tail: maps excerpt, RSS growth, fault deltas.
func report(addr, length uintptr, mode string, base baseline) error {
	if err := printMapsExcerpt(addr, length, mode); err != nil {
		return err
	}
	rssAfter, err := vmrssKB()
	if err != nil {
		return err
	}
	minor, major, err := faultCounts()
	if err != nil {
		return err
	}
	fmt.Printf("memmap: vmrss_before=%dKB vmrss_after=%dKB\n", base.rssKB, rssAfter)
	fmt.Printf("memmap: faults minor=%d major=%d\n", minor-base.minor, major-base.major)
	return nil
}

// ---------------------------------------------------------------------------
// Modes
// ---------------------------------------------------------------------------

func touchWritable(b []byte, page int) {
	for i := 0; i < len(b); i += page {
		b[i] = 1
	}
	if len(b) > 0 {
		sink = b[0]
	}
}

func touchReadable(b []byte, page int) {
	var sum byte
	for i := 0; i < len(b); i += page {
		sum += b[i]
	}
	sink = sum
}

// runStack touches a large local buffer on a dedicated goroutine so the
// growth happens on a fresh stack; the result comes back over a channel.
// The baseline is captured before the goroutine starts so the runtime work
// of growing (or relocating) that stack is part of the measured window.
func runStack(bytes, page int) error {
	base, err := takeBaseline()
	if err != nil {
		return err
	}
	errc := make(chan error, 1)
	go func() {
		errc <- stackWorker(bytes, page, base)
	}()
	return <-errc
}

//go:noinline
func stackWorker(n, page int, base baseline) error {
	var buf [stackCapMb * mib]byte
	touchWritable(buf[:n], page)
	return report(uintptr(unsafe.Pointer(&buf[0])), uintptr(n), "stack", base)
}

func runHeap(n, page int) error {
	base, err := takeBaseline()
	if err != nil {
		return err
	}
	buf := make([]byte, n)
	touchWritable(buf, page)
	return report(uintptr(unsafe.Pointer(&buf[0])), uintptr(n), "heap", base)
}

func runMmapAnon(n, page int) error {
	base, err := takeBaseline()
	if err != nil {
		return err
	}
	m, err := unix.Mmap(-1, 0, n, unix.PROT_READ|unix.PROT_WRITE,
		unix.MAP_PRIVATE|unix.MAP_ANONYMOUS)
	if err != nil {
		return fmt.Errorf("mmap: %w", err)
	}
	defer unix.Munmap(m)
	touchWritable(m, page)
	return report(uintptr(unsafe.Pointer(&m[0])), uintptr(n), "mmap-anon", base)
}

func openForMap(path string) (*os.File, int64, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, 0, err
	}
	st, err := f.Stat()
	if err != nil {
		f.Close()
		return nil, 0, fmt.Errorf("%s: stat: %w", path, err)
	}
	if st.Size() <= 0 {
		f.Close()
		return nil, 0, fmt.Errorf("%s: file is empty", path)
	}
	return f, st.Size(), nil
}

func runMmapFile(path string, page int) error {
	f, size, err := openForMap(path)
	if err != nil {
		return err
	}
	defer f.Close()
	pages := (size + int64(page) - 1) / int64(page)
	fmt.Printf("memmap: mode=mmap-file bytes=%d pages=%d\n", size, pages)
	base, err := takeBaseline()
	if err != nil {
		return err
	}
	m, err := unix.Mmap(int(f.Fd()), 0, int(size), unix.PROT_READ, unix.MAP_PRIVATE)
	if err != nil {
		return fmt.Errorf("mmap: %w", err)
	}
	defer unix.Munmap(m)
	touchReadable(m, page)
	return report(uintptr(unsafe.Pointer(&m[0])), uintptr(size), "mmap-file", base)
}

func writeWalkFile(n int) (string, error) {
	path := filepath.Join(os.TempDir(),
		fmt.Sprintf("memmap-walk-%d.bin", os.Getpid()))
	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
	if err != nil {
		return "", err
	}
	defer f.Close()
	chunk := make([]byte, mib)
	for i := range chunk {
		chunk[i] = 0xA5
	}
	for left := n; left > 0; left -= len(chunk) {
		c := chunk
		if left < len(c) {
			c = c[:left]
		}
		if _, err := f.Write(c); err != nil {
			return "", fmt.Errorf("%s: write: %w", path, err)
		}
	}
	return path, nil
}

func runFaultWalk(n, page int) error {
	path, err := writeWalkFile(n)
	if err != nil {
		return err
	}
	defer os.Remove(path)
	f, size, err := openForMap(path)
	if err != nil {
		return err
	}
	defer f.Close()
	base, err := takeBaseline()
	if err != nil {
		return err
	}
	m, err := unix.Mmap(int(f.Fd()), 0, int(size), unix.PROT_READ, unix.MAP_PRIVATE)
	if err != nil {
		return fmt.Errorf("mmap: %w", err)
	}
	defer unix.Munmap(m)

	fmt.Printf("memmap: walk file=%s steps=%d\n", path, walkSteps)
	pages := int(size) / page
	prevMinor, prevMajor := base.minor, base.major
	done := 0
	for step := 1; step <= walkSteps; step++ {
		quota := pages / walkSteps
		if step == walkSteps {
			quota = pages - done
		}
		touchReadable(m[done*page:(done+quota)*page], page)
		minor, major, err := faultCounts()
		if err != nil {
			return err
		}
		fmt.Printf("memmap: step=%d/%d pages=%d minor=%d major=%d\n",
			step, walkSteps, quota, minor-prevMinor, major-prevMajor)
		prevMinor, prevMajor = minor, major
		done += quota
	}
	return report(uintptr(unsafe.Pointer(&m[0])), uintptr(size), "fault-walk", base)
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

func usage() {
	fmt.Fprintln(os.Stderr,
		"usage: memmap --mode stack|heap|mmap-anon|mmap-file <FILE>|fault-walk [--mb N]")
}

func parseArgs(args []string) (options, error) {
	opts := options{mb: 64}
	modeSet := false
	for i := 0; i < len(args); i++ {
		switch a := args[i]; {
		case a == "--mode":
			i++
			if i >= len(args) {
				return opts, fmt.Errorf("--mode needs a value")
			}
			switch args[i] {
			case "stack", "heap", "mmap-anon", "mmap-file", "fault-walk":
				opts.mode = args[i]
				modeSet = true
			default:
				return opts, fmt.Errorf("unknown mode: %s", args[i])
			}
		case a == "--mb":
			i++
			if i >= len(args) {
				return opts, fmt.Errorf("--mb needs a value")
			}
			mb, err := strconv.Atoi(args[i])
			if err != nil || mb < 1 || mb > 1024 {
				return opts, fmt.Errorf("--mb must be 1..1024")
			}
			opts.mb = mb
		case !strings.HasPrefix(a, "--") && opts.file == "":
			opts.file = a
		default:
			return opts, fmt.Errorf("unexpected argument: %s", a)
		}
	}
	if !modeSet {
		return opts, fmt.Errorf("--mode is required")
	}
	if opts.mode == "mmap-file" && opts.file == "" {
		return opts, fmt.Errorf("mmap-file needs a FILE argument")
	}
	if opts.mode != "mmap-file" && opts.file != "" {
		return opts, fmt.Errorf("only mmap-file takes a FILE argument")
	}
	return opts, nil
}

func main() {
	opts, err := parseArgs(os.Args[1:])
	if err != nil {
		usage()
		os.Exit(2)
	}
	page := unix.Getpagesize()

	switch opts.mode {
	case "stack":
		n := min(opts.mb, stackCapMb) * mib
		fmt.Printf("memmap: mode=stack bytes=%d pages=%d\n", n, n/page)
		err = runStack(n, page)
	case "heap":
		n := opts.mb * mib
		fmt.Printf("memmap: mode=heap bytes=%d pages=%d\n", n, n/page)
		err = runHeap(n, page)
	case "mmap-anon":
		n := opts.mb * mib
		fmt.Printf("memmap: mode=mmap-anon bytes=%d pages=%d\n", n, n/page)
		err = runMmapAnon(n, page)
	case "mmap-file":
		err = runMmapFile(opts.file, page) // prints its own header line
	case "fault-walk":
		n := opts.mb * mib
		fmt.Printf("memmap: mode=fault-walk bytes=%d pages=%d\n", n, n/page)
		err = runFaultWalk(n, page)
	}
	if err != nil {
		fmt.Fprintf(os.Stderr, "memmap: error: %v\n", err)
		os.Exit(1)
	}
}

File: examples/15-virtual-memory/cpp/src/main.cpp
// memmap: self-inspecting virtual-memory tool — perform one kind of
// allocation, touch its pages, then report what the kernel now shows for
// this very process: the /proc/self/maps region backing the allocation,
// the VmRSS growth, and the getrusage(2) page-fault deltas.
//
//   memmap --mode stack|heap|mmap-anon|mmap-file <FILE>|fault-walk [--mb N]
//
//   stack      touch a local buffer on the call stack (clamped to 4 MiB)
//   heap       allocator memory (operator new via std::vector)
//   mmap-anon  private anonymous mmap(2), written one byte per page
//   mmap-file  read-only private mmap of FILE, read one byte per page
//   fault-walk create a temp file of N MiB, map it, touch it in 8 steps,
//              printing the minor/major fault growth after every step
//
// Output contract (identical across C++, Go, Rust):
//   memmap: mode=<m> bytes=<b> pages=<p>
//   [fault-walk only] memmap: walk file=<path> steps=8
//   [fault-walk only] memmap: step=<i>/8 pages=<p> minor=<d> major=<d>
//   memmap: maps excerpt
//   memmap:   <raw /proc/self/maps line>   <-- target (mode=<m>)
//   memmap: vmrss_before=<kb>KB vmrss_after=<kb>KB
//   memmap: faults minor=<n> major=<n>

#include <array>
#include <cerrno>
#include <charconv>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <expected>
#include <fstream>
#include <print>
#include <span>
#include <string>
#include <string_view>
#include <system_error>
#include <utility>
#include <vector>

#include <fcntl.h>
#include <sys/mman.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <unistd.h>

namespace {

constexpr std::size_t kMiB = 1024 * 1024;
constexpr std::size_t kStackCapMb = 4;  // stays well inside the 8 MiB rlimit
constexpr int kWalkSteps = 8;

enum class Mode { stack, heap, mmap_anon, mmap_file, fault_walk };

[[nodiscard]] constexpr std::string_view mode_name(Mode m) {
    switch (m) {
    case Mode::stack:      return "stack";
    case Mode::heap:       return "heap";
    case Mode::mmap_anon:  return "mmap-anon";
    case Mode::mmap_file:  return "mmap-file";
    case Mode::fault_walk: return "fault-walk";
    }
    return "?";
}

[[nodiscard]] std::string errno_message(int err) {
    return std::error_code(err, std::system_category()).message();
}

// Keep the optimizer from proving the touched buffer unread.
inline void opt_barrier(const void* p) {
    __asm__ volatile("" : : "r"(p) : "memory");
}

// ---------------------------------------------------------------------------
// RAII wrappers
// ---------------------------------------------------------------------------

class Fd {
public:
    explicit Fd(int fd = -1) noexcept : fd_{fd} {}
    Fd(const Fd&) = delete;
    Fd& operator=(const Fd&) = delete;
    Fd(Fd&& o) noexcept : fd_{std::exchange(o.fd_, -1)} {}
    Fd& operator=(Fd&& o) noexcept {
        if (this != &o) {
            reset();
            fd_ = std::exchange(o.fd_, -1);
        }
        return *this;
    }
    ~Fd() { reset(); }
    [[nodiscard]] int get() const noexcept { return fd_; }
    [[nodiscard]] bool valid() const noexcept { return fd_ >= 0; }

private:
    void reset() noexcept {
        if (fd_ >= 0) {
            ::close(fd_);
            fd_ = -1;
        }
    }
    int fd_ = -1;
};

class Mapping {
public:
    Mapping(void* p, std::size_t len) noexcept : p_{p}, len_{len} {}
    Mapping(const Mapping&) = delete;
    Mapping& operator=(const Mapping&) = delete;
    Mapping(Mapping&& o) noexcept
        : p_{std::exchange(o.p_, nullptr)}, len_{std::exchange(o.len_, 0)} {}
    ~Mapping() {
        if (p_ != nullptr) {
            ::munmap(p_, len_);
        }
    }
    [[nodiscard]] std::byte* data() const noexcept {
        return static_cast<std::byte*>(p_);
    }
    [[nodiscard]] std::size_t size() const noexcept { return len_; }

private:
    void* p_ = nullptr;
    std::size_t len_ = 0;
};

// Unlinks its path on scope exit.
class TempFile {
public:
    explicit TempFile(std::string path) : path_{std::move(path)} {}
    TempFile(const TempFile&) = delete;
    TempFile& operator=(const TempFile&) = delete;
    ~TempFile() {
        if (!path_.empty()) {
            ::unlink(path_.c_str());
        }
    }
    [[nodiscard]] const std::string& path() const noexcept { return path_; }

private:
    std::string path_;
};

// ---------------------------------------------------------------------------
// /proc and getrusage probes
// ---------------------------------------------------------------------------

[[nodiscard]] std::expected<long, std::string> vmrss_kb() {
    std::ifstream f{"/proc/self/status"};
    if (!f) {
        return std::unexpected{"cannot open /proc/self/status"};
    }
    std::string line;
    while (std::getline(f, line)) {
        if (line.starts_with("VmRSS:")) {
            std::string_view rest{line};
            rest.remove_prefix(6);
            std::size_t i = 0;
            while (i < rest.size() && (rest[i] == ' ' || rest[i] == '\t')) ++i;
            long kb = 0;
            auto [ptr, ec] =
                std::from_chars(rest.data() + i, rest.data() + rest.size(), kb);
            if (ec != std::errc{}) {
                return std::unexpected{"cannot parse VmRSS"};
            }
            return kb;
        }
    }
    return std::unexpected{"VmRSS not found in /proc/self/status"};
}

struct Faults {
    long minor = 0;
    long major = 0;
};

[[nodiscard]] std::expected<Faults, std::string> fault_counts() {
    rusage ru{};
    if (::getrusage(RUSAGE_SELF, &ru) != 0) {
        return std::unexpected{"getrusage: " + errno_message(errno)};
    }
    return Faults{ru.ru_minflt, ru.ru_majflt};
}

// Print every /proc/self/maps line whose range overlaps [addr, addr+len).
[[nodiscard]] std::expected<void, std::string>
print_maps_excerpt(std::uintptr_t addr, std::size_t len, Mode mode) {
    std::ifstream f{"/proc/self/maps"};
    if (!f) {
        return std::unexpected{"cannot open /proc/self/maps"};
    }
    std::println("memmap: maps excerpt");
    const std::uintptr_t lo = addr;
    const std::uintptr_t hi = addr + len;
    std::string line;
    while (std::getline(f, line)) {
        const auto dash = line.find('-');
        const auto sp = line.find(' ');
        if (dash == std::string::npos || sp == std::string::npos || sp < dash) {
            continue;
        }
        std::uintptr_t start = 0;
        std::uintptr_t end = 0;
        auto r1 = std::from_chars(line.data(), line.data() + dash, start, 16);
        auto r2 =
            std::from_chars(line.data() + dash + 1, line.data() + sp, end, 16);
        if (r1.ec != std::errc{} || r2.ec != std::errc{}) {
            continue;
        }
        if (start < hi && end > lo) {
            std::println("memmap:   {}   <-- target (mode={})", line,
                         mode_name(mode));
        }
    }
    return {};
}

struct Baseline {
    long rss_kb = 0;
    Faults faults;
};

[[nodiscard]] std::expected<Baseline, std::string> take_baseline() {
    auto rss = vmrss_kb();
    if (!rss) return std::unexpected{rss.error()};
    auto fl = fault_counts();
    if (!fl) return std::unexpected{fl.error()};
    return Baseline{*rss, *fl};
}

// Common tail: maps excerpt, RSS growth, fault deltas.
[[nodiscard]] std::expected<void, std::string>
report(std::uintptr_t addr, std::size_t len, Mode mode, const Baseline& base) {
    if (auto r = print_maps_excerpt(addr, len, mode); !r) return r;
    auto rss_after = vmrss_kb();
    if (!rss_after) return std::unexpected{rss_after.error()};
    auto fl = fault_counts();
    if (!fl) return std::unexpected{fl.error()};
    std::println("memmap: vmrss_before={}KB vmrss_after={}KB", base.rss_kb,
                 *rss_after);
    std::println("memmap: faults minor={} major={}",
                 fl->minor - base.faults.minor, fl->major - base.faults.major);
    return {};
}

// ---------------------------------------------------------------------------
// Modes
// ---------------------------------------------------------------------------

void touch_writable(std::span<std::byte> mem, std::size_t page) {
    for (std::size_t i = 0; i < mem.size(); i += page) {
        mem[i] = std::byte{1};
    }
    opt_barrier(mem.data());
}

// Read one byte per page; the sum flows into the barrier so it stays live.
void touch_readable(std::span<const std::byte> mem, std::size_t page) {
    unsigned long sum = 0;
    for (std::size_t i = 0; i < mem.size(); i += page) {
        sum += std::to_underlying(mem[i]);
    }
    opt_barrier(&sum);
}

// Baseline is captured by the caller so that any page the compiler touches
// while setting up this 4 MiB frame still lands inside the measured window.
[[gnu::noinline]] std::expected<void, std::string>
stack_worker(std::size_t bytes, std::size_t page, const Baseline& base) {
    std::array<std::byte, kStackCapMb * kMiB> buf{};
    touch_writable(std::span{buf.data(), bytes}, page);
    return report(reinterpret_cast<std::uintptr_t>(buf.data()), bytes,
                  Mode::stack, base);
}

[[nodiscard]] std::expected<void, std::string>
run_stack(std::size_t bytes, std::size_t page) {
    auto base = take_baseline();
    if (!base) return std::unexpected{base.error()};
    return stack_worker(bytes, page, *base);
}

[[nodiscard]] std::expected<void, std::string>
run_heap(std::size_t bytes, std::size_t page) {
    auto base = take_baseline();
    if (!base) return std::unexpected{base.error()};
    std::vector<std::byte> buf(bytes);
    touch_writable(std::span{buf}, page);
    return report(reinterpret_cast<std::uintptr_t>(buf.data()), bytes,
                  Mode::heap, *base);
}

[[nodiscard]] std::expected<void, std::string>
run_mmap_anon(std::size_t bytes, std::size_t page) {
    auto base = take_baseline();
    if (!base) return std::unexpected{base.error()};
    void* p = ::mmap(nullptr, bytes, PROT_READ | PROT_WRITE,
                     MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (p == MAP_FAILED) {
        return std::unexpected{"mmap: " + errno_message(errno)};
    }
    Mapping map{p, bytes};
    touch_writable(std::span{map.data(), map.size()}, page);
    return report(reinterpret_cast<std::uintptr_t>(map.data()), map.size(),
                  Mode::mmap_anon, *base);
}

[[nodiscard]] std::expected<std::pair<Fd, std::size_t>, std::string>
open_for_map(const std::string& path) {
    Fd fd{::open(path.c_str(), O_RDONLY | O_CLOEXEC)};
    if (!fd.valid()) {
        return std::unexpected{path + ": " + errno_message(errno)};
    }
    struct stat st{};
    if (::fstat(fd.get(), &st) != 0) {
        return std::unexpected{path + ": fstat: " + errno_message(errno)};
    }
    if (st.st_size <= 0) {
        return std::unexpected{path + ": file is empty"};
    }
    return std::pair{std::move(fd), static_cast<std::size_t>(st.st_size)};
}

[[nodiscard]] std::expected<Mapping, std::string> map_readonly(int fd,
                                                               std::size_t len) {
    void* p = ::mmap(nullptr, len, PROT_READ, MAP_PRIVATE, fd, 0);
    if (p == MAP_FAILED) {
        return std::unexpected{"mmap: " + errno_message(errno)};
    }
    return Mapping{p, len};
}

[[nodiscard]] std::expected<void, std::string>
run_mmap_file(const std::string& path, std::size_t page) {
    auto opened = open_for_map(path);
    if (!opened) return std::unexpected{opened.error()};
    auto& [fd, len] = *opened;
    const std::size_t pages = (len + page - 1) / page;
    std::println("memmap: mode=mmap-file bytes={} pages={}", len, pages);
    auto base = take_baseline();
    if (!base) return std::unexpected{base.error()};
    auto map = map_readonly(fd.get(), len);
    if (!map) return std::unexpected{map.error()};
    touch_readable(std::span{map->data(), map->size()}, page);
    return report(reinterpret_cast<std::uintptr_t>(map->data()), map->size(),
                  Mode::mmap_file, *base);
}

[[nodiscard]] std::expected<std::string, std::string>
write_walk_file(std::size_t bytes) {
    const char* tmpdir = std::getenv("TMPDIR");
    std::string path = std::string{tmpdir != nullptr ? tmpdir : "/tmp"} +
                       "/memmap-walk-" + std::to_string(::getpid()) + ".bin";
    Fd fd{::open(path.c_str(), O_WRONLY | O_CREAT | O_TRUNC | O_CLOEXEC, 0644)};
    if (!fd.valid()) {
        return std::unexpected{path + ": " + errno_message(errno)};
    }
    std::vector<std::byte> chunk(kMiB, std::byte{0xA5});
    std::size_t left = bytes;
    while (left > 0) {
        const std::size_t n = std::min(left, chunk.size());
        std::size_t off = 0;
        while (off < n) {
            const ssize_t w = ::write(fd.get(), chunk.data() + off, n - off);
            if (w < 0) {
                if (errno == EINTR) continue;
                return std::unexpected{path + ": write: " +
                                       errno_message(errno)};
            }
            off += static_cast<std::size_t>(w);
        }
        left -= n;
    }
    return path;
}

[[nodiscard]] std::expected<void, std::string>
run_fault_walk(std::size_t bytes, std::size_t page) {
    auto path = write_walk_file(bytes);
    if (!path) return std::unexpected{path.error()};
    TempFile tmp{*path};
    auto opened = open_for_map(tmp.path());
    if (!opened) return std::unexpected{opened.error()};
    auto& [fd, len] = *opened;
    auto base = take_baseline();
    if (!base) return std::unexpected{base.error()};
    auto map = map_readonly(fd.get(), len);
    if (!map) return std::unexpected{map.error()};

    std::println("memmap: walk file={} steps={}", tmp.path(), kWalkSteps);
    const std::size_t pages = len / page;
    Faults prev = base->faults;
    std::size_t done = 0;
    for (int step = 1; step <= kWalkSteps; ++step) {
        std::size_t quota = pages / kWalkSteps;
        if (step == kWalkSteps) quota = pages - done;
        touch_readable(std::span{map->data() + done * page, quota * page},
                       page);
        auto now = fault_counts();
        if (!now) return std::unexpected{now.error()};
        std::println("memmap: step={}/{} pages={} minor={} major={}", step,
                     kWalkSteps, quota, now->minor - prev.minor,
                     now->major - prev.major);
        prev = *now;
        done += quota;
    }
    return report(reinterpret_cast<std::uintptr_t>(map->data()), map->size(),
                  Mode::fault_walk, *base);
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

struct Options {
    Mode mode = Mode::heap;
    std::size_t mb = 64;
    std::string file;
};

void usage(std::FILE* to) {
    std::println(to,
                 "usage: memmap --mode stack|heap|mmap-anon|mmap-file <FILE>|"
                 "fault-walk [--mb N]");
}

[[nodiscard]] std::expected<Options, std::string>
parse_args(std::span<char*> args) {
    Options opts;
    bool mode_set = false;
    for (std::size_t i = 0; i < args.size(); ++i) {
        const std::string_view a{args[i]};
        if (a == "--mode") {
            if (++i >= args.size()) {
                return std::unexpected{"--mode needs a value"};
            }
            const std::string_view m{args[i]};
            if (m == "stack")           opts.mode = Mode::stack;
            else if (m == "heap")       opts.mode = Mode::heap;
            else if (m == "mmap-anon")  opts.mode = Mode::mmap_anon;
            else if (m == "mmap-file")  opts.mode = Mode::mmap_file;
            else if (m == "fault-walk") opts.mode = Mode::fault_walk;
            else return std::unexpected{"unknown mode: " + std::string{m}};
            mode_set = true;
        } else if (a == "--mb") {
            if (++i >= args.size()) {
                return std::unexpected{"--mb needs a value"};
            }
            const std::string_view v{args[i]};
            std::size_t mb = 0;
            auto [p, ec] = std::from_chars(v.data(), v.data() + v.size(), mb);
            if (ec != std::errc{} || p != v.data() + v.size() || mb < 1 ||
                mb > 1024) {
                return std::unexpected{"--mb must be 1..1024"};
            }
            opts.mb = mb;
        } else if (!a.starts_with("--") && opts.file.empty()) {
            opts.file = a;
        } else {
            return std::unexpected{"unexpected argument: " + std::string{a}};
        }
    }
    if (!mode_set) return std::unexpected{"--mode is required"};
    if (opts.mode == Mode::mmap_file && opts.file.empty()) {
        return std::unexpected{"mmap-file needs a FILE argument"};
    }
    if (opts.mode != Mode::mmap_file && !opts.file.empty()) {
        return std::unexpected{"only mmap-file takes a FILE argument"};
    }
    return opts;
}

}  // namespace

int main(int argc, char** argv) {
    const auto opts =
        parse_args(std::span{argv + 1, static_cast<std::size_t>(argc - 1)});
    if (!opts) {
        usage(stderr);
        return 2;
    }
    const auto page = static_cast<std::size_t>(::sysconf(_SC_PAGESIZE));

    std::expected<void, std::string> r;
    switch (opts->mode) {
    case Mode::stack: {
        const std::size_t bytes = std::min(opts->mb, kStackCapMb) * kMiB;
        std::println("memmap: mode=stack bytes={} pages={}", bytes,
                     bytes / page);
        r = run_stack(bytes, page);
        break;
    }
    case Mode::heap: {
        const std::size_t bytes = opts->mb * kMiB;
        std::println("memmap: mode=heap bytes={} pages={}", bytes,
                     bytes / page);
        r = run_heap(bytes, page);
        break;
    }
    case Mode::mmap_anon: {
        const std::size_t bytes = opts->mb * kMiB;
        std::println("memmap: mode=mmap-anon bytes={} pages={}", bytes,
                     bytes / page);
        r = run_mmap_anon(bytes, page);
        break;
    }
    case Mode::mmap_file:
        r = run_mmap_file(opts->file, page);  // prints its own header line
        break;
    case Mode::fault_walk: {
        const std::size_t bytes = opts->mb * kMiB;
        std::println("memmap: mode=fault-walk bytes={} pages={}", bytes,
                     bytes / page);
        r = run_fault_walk(bytes, page);
        break;
    }
    }
    if (!r) {
        std::println(stderr, "memmap: error: {}", r.error());
        return 1;
    }
    return 0;
}
